config/initializers/vault.rb

```diff
@@ -1,50 +1,72 @@
-VAULT_ENABLED = ENV.fetch('SECRET_STORAGE_BACKEND', false)
-if VAULT_ENABLED == 'SecretStorage::HashicorpVault'
+if ENV["SECRET_STORAGE_BACKEND"] == "SecretStorage::HashicorpVault"
   require 'vault'
   Rails.logger.info("Vault Client enabled")
-  Vault.configure do |config|
-    config.ssl_pem_file = ENV.fetch("VAULT_SSL_CERT")
-    config.ssl_verify = ActiveRecord::Type::Boolean.new.type_cast_from_user(ENV.fetch("VAULT_SSL_VERIFY", true))
-    config.address = ENV.fetch("VAULT_ADDR", 'https://127.0.0.1:8200')
-
-    # Timeout the connection after a certain amount of time (seconds)
-    config.timeout = 5
-
-    # It is also possible to have finer-grained controls over the timeouts, these
-    # may also be read as environment variables
-    config.ssl_timeout  = 3
-    config.open_timeout = 3
-    config.read_timeout = 2
-  end
 
-  # build our own client here so we can write to multiple
-  # vaults based on env
   class VaultClient < Vault::Client
     PEM = File.read(Vault.ssl_pem_file)
     CERT_AUTH_PATH = '/v1/auth/cert/login'.freeze
     DEFAULT_CLIENT_OPTIONS = {
       use_ssl: true,
+      ssl_pem_file: Rails.root.join(ENV.fetch("VAULT_SSL_CERT")),
       verify_mode: ENV.fetch("VAULT_SSL_VERIFY", 1).to_i,
+      timeout: 5,
+      ssl_timeout: 3,
+      open_timeout: 3,
+      read_timeout: 2,
       cert: OpenSSL::X509::Certificate.new(PEM),
       key: OpenSSL::PKey::RSA.new(PEM)
     }.freeze
 
     def initialize
-      # we are just stubbing in our own auth, everything else should fall
-      # through to our parrent
-      super
-
-      # if we are testing, just return here.  We'll let super configure
-      # the rest of the client
       return if Rails.env.test?
-      uri = URI.parse(Vault.address)
-      @http = Net::HTTP.start(uri.host, uri.port, DEFAULT_CLIENT_OPTIONS)
-      response = @http.request(Net::HTTP::Post.new(CERT_AUTH_PATH))
-      if response.code == "200"
-        @token = JSON.parse(response.body).fetch("auth")["client_token"]
-      else
-        raise "Failed to get auth token from vault server"
+      @writers = []
+      # since we have a bunch of servers, don't use the client singlton to
+      # talk to them
+      vault_hosts.each_with_index do |vault_server, idx|
+        @writers[idx] = Vault::Client.new(DEFAULT_CLIENT_OPTIONS)
+        @writers[idx].address = vault_server
+
+        uri = URI.parse(vault_server)
+        @http = Net::HTTP.start(uri.host, uri.port, DEFAULT_CLIENT_OPTIONS)
```

Comment:
this seems counter intuitive ... call some method on the writer instead of re-implementing connection logic ?